src/lib/operators/jit_operator/jit_operations.hpp

```diff
@@ -0,0 +1,207 @@
+ * template substitution if they cannot perform their operation on a given combination of input types. In this case,
+ * the "InvalidTypeCatcher" provides a default implementation that throws an exception.
+ *
+ * The generic lambdas can also be passed to the "jit_compute_type" function. The function uses the same dispatching
+ * mechanisms. But instead of executing a computation, it only determines the result type the computation would
+ * have if it were carried out. This functionality is used to determine the type of intermediate values and
+ * computed output columns.
+ *
+ * Logical operators, IsNull and IsNotNull must be handled separately, since their NULL value semantics are
+ * different (i.e. a NULL as either input does not result in the output being NULL as well).
+ */
+
+#define JIT_GET_ENUM_VALUE(index, s) APPEND_ENUM_NAMESPACE(_, _, BOOST_PP_TUPLE_ELEM(3, 1, BOOST_PP_SEQ_ELEM(index, s)))
+#define JIT_GET_DATA_TYPE(index, s) BOOST_PP_TUPLE_ELEM(3, 0, BOOST_PP_SEQ_ELEM(index, s))
+
+#define JIT_COMPUTE_CASE(r, types)                                                                                   \
+  case static_cast<uint8_t>(JIT_GET_ENUM_VALUE(0, types)) << 8 | static_cast<uint8_t>(JIT_GET_ENUM_VALUE(1, types)): \
+    catching_func(lhs.get<JIT_GET_DATA_TYPE(0, types)>(), rhs.get<JIT_GET_DATA_TYPE(1, types)>(), result);           \
+    break;
+
+#define JIT_COMPUTE_TYPE_CASE(r, types)                                                                              \
+  case static_cast<uint8_t>(JIT_GET_ENUM_VALUE(0, types)) << 8 | static_cast<uint8_t>(JIT_GET_ENUM_VALUE(1, types)): \
+    return catching_func(JIT_GET_DATA_TYPE(0, types)(), JIT_GET_DATA_TYPE(1, types)());
+
+/* Arithmetic operators */
+const auto jit_addition = [](const auto a, const auto b) -> decltype(a + b) { return a + b; };
+const auto jit_subtraction = [](const auto a, const auto b) -> decltype(a - b) { return a - b; };
+const auto jit_multiplication = [](const auto a, const auto b) -> decltype(a * b) { return a * b; };
+const auto jit_division = [](const auto a, const auto b) -> decltype(a / b) { return a / b; };
+const auto jit_modulo = [](const auto a, const auto b) -> decltype(a % b) { return a % b; };
+const auto jit_power = [](const auto a, const auto b) -> decltype(std::pow(a, b)) { return std::pow(a, b); };
+
+/* Comparison operators */
+const auto jit_equals = [](const auto a, const auto b) -> decltype(a == b) { return a == b; };
+const auto jit_not_equals = [](const auto a, const auto b) -> decltype(a != b) { return a != b; };
+const auto jit_less_than = [](const auto a, const auto b) -> decltype(a < b) { return a < b; };
+const auto jit_less_than_equals = [](const auto a, const auto b) -> decltype(a <= b) { return a <= b; };
+const auto jit_greater_than = [](const auto a, const auto b) -> decltype(a > b) { return a > b; };
+const auto jit_greater_than_equals = [](const auto a, const auto b) -> decltype(a >= b) { return a >= b; };
```

Comment:
Do we need the decltypes here?

Reply:
Waiting for the compiler ;)
I am pretty sure we need them here for the SFINAE pattern to work. But I am testing this right now

Reply:
Ok. I thought you missed it because all the other points were addressed.

Reply:
Compiler says we need them.